fix: json format output with multiline content

pyxide opened the pull request:
Hi,

I have added a function `toJsonString` to handle multiline content and escape special characters (quote, backslash) that would break the JSON output.

I’m using it for the `subject` and `body` fields that broke in my case, it could be used for other fields as well.
- `%n__START_STRING__body%n%b%n__END_STRING__,%n` → `"body": "{escaped body}",`
- `%n__START_STRING__subject%n%b%n__END_STRING__,%n` → `"subject": "{escaped subject}",`

tomice commented:
Hello, and thank you for your contribution! I have a couple things I'd like to note, if you don't mind.

* Normally, when we create functions, we like to outline them with the `function` keyword before them. While it isn't technically needed in bash, it helps with large single file codebases like this, and it helps with consistency since all of the other functions have the `function` keyword prepended to them.
* In your comments, specifically in the `OUTS:` (line 533), it mentions that this returns a JSON formatted file. I don't believe that is the case for this function, correct? I believe this is a helper function that exists solely to format a provided string and print it to `stdout` instead out passing data around for another function to consume. In our case, if a function only ever prints to `stdout`, then we set `OUTS` to `None`. Let me know if I'm incorrect here, though.
* Also, for consistency, we like to provide an `ARGS` in the comments whether it takes one or not. In this instance, I believe it does indeed take an argument which would be a string in which the helper function performs its JSON translation, correct?
* Can you provide an example codebase where multiline comments were broken in the current version and where this commit fixes that issue?

pyxide commented:
Thank you for your thorough review and your amenity.

> Normally, when we create functions, we like to outline them with the function keyword

I have added the function keyword.

> it mentions that this returns a JSON formatted file. I don't believe that is the case for this function, correct?

Yes, you are right, I have renamed it to `toJsonProp` and modified the documentation accordingly.

> we like to provide an ARGS in the comments

There is now one optional property `propTag` and it’s documented.

> Can you provide an example codebase

While writing the test, I have noticed some errors and refactored code.
- the new function was renamed from `toJsonString` to `toJsonProp`.
- the `commit_notes` field is also escaped.
- the property tag is now randomized to avoid collision with the actual content: `__JSONPROP${RANDOM}__`
- tabulation are escaped. In fact, all the characters between `\0x00` and `0x1F` should be escaped . 
- I have split the long git formatting string in many lines.

```
#rm -Rf test-git/.git
mkdir test-git
cd test-git
git init
printf '\n[user]\nname = test-git\nemail = test-git@example.org\n'> .git/config
printf 'test-git\n========\n' > README.txt
git add README.txt
git commit -m 'added readme (o\w/o)' -m 'in markdown, no \r\n, only \n' \
  -m 'a very "simple" readme'
git notes add -m 'Some notes' -m 'And another one'
git log
```

If you run the actual `git-quick-stats -j`, you will get an invalid JSON content:

```txt
$ cat output.json
[
{
  "commit": "5a87dd0c296ab36ac4c2fd73d04f31f91c3cef11",
  "abbreviated_commit": "5a87dd0",
  "tree": "937aa44d15611c2ae0214e187c3370048d92abad",
  "abbreviated_tree": "937aa44",
  "parent": "",
  "abbreviated_parent": "",
  "refs": "HEAD -> master",
  "encoding": "",
  "subject": "added readme (o\w/o)",
  "sanitized_subject_line": "added-readme-o-w-o",
  "body": "in markdown, no \r\n, only \n

a very "simple" readme
",
  "commit_notes": "Some notes

And another one
",
  "author": {
    "name": "test-git",
    "email": "test-git@example.org",
    "date": "Wed, 6 Jul 2022 22:25:29 +0200"
  },
  "commiter": {
    "name": "test-git",
    "email": "test-git@example.org",
    "date": "Wed, 6 Jul 2022 22:25:29 +0200"
  }
}
]
```
When you run the patched code, you get a correct JSON file
```json
[
{
  "commit": "5a87dd0c296ab36ac4c2fd73d04f31f91c3cef11",
  "abbreviated_commit": "5a87dd0",
  "tree": "937aa44d15611c2ae0214e187c3370048d92abad",
  "abbreviated_tree": "937aa44",
  "parent": "",
  "abbreviated_parent": "",
  "refs": "HEAD -> master",
  "encoding": "",
  "subject": "added readme (o\\w/o)",
  "sanitized_subject_line": "added-readme-o-w-o",
  "body": "in markdown, no \\r\\n, only \\n\n\na very \"simple\" readme\n",
  "commit_notes": "Some notes\n\nAnd another one\n",
  "author": {
    "name": "test-git",
    "email": "test-git@example.org",
    "date": "Wed, 6 Jul 2022 22:25:29 +0200"
  },
  "commiter": {
    "name": "test-git",
    "email": "test-git@example.org",
    "date": "Wed, 6 Jul 2022 22:25:29 +0200"
  }
}
]
```

tomice commented:
Thanks for providing examples of the JSON output's current code being broken, as well as showing the new PR's code fixing the issue. That helps a ton!

I see that you found a slight issue with your original implementation regarding handling characters `0x00` to `0x1F`. Does this also work with characters `0x21` to `0x2F`, `0x3A` to `0x40`, `0x5B` to `0x60`, and `0x7B` to `0x7E`? 

Link for ease of reference: https://man7.org/linux/man-pages/man7/ascii.7.html

pyxide commented:
Hi,

I have added the JSON encoding for ASCII control chars, and added the test script in `tests/test-git/resetgit`.
Everytime you run it from its directory, it erases the last project if any and it generates a new Git project with a commit.

I have added the author and commiter name and email to the `toJsonProp` process, see the test script.

tomice commented:
Really appreciate the extra test script. This could be a useful way for us to do some internal testing if we need it in the future. Everything looks solid on my end. Thanks so much for the PR!
